stream_consumer.go

```diff
@@ -85,57 +90,179 @@ func NewStreamConsumer(streamName string, endpointType EndpointType, endpoints [
 	r.InitialAddrs(addresses)
 	target := r.Scheme() + ":///stream"
 
+	config := defaultStreamEndpointConfig()
+	for _, opt := range opts {
+		opt(config)
+	}
+	conn, err := grpc.Dial(target, grpc.WithInsecure(), grpc.WithBalancerName(roundrobin.Name), grpc.WithInsecure(),
+		grpc.WithDefaultCallOptions(grpc.ForceCodec(&gogoCodec{})),
+		grpc.WithBackoffMaxDelay(config.backoffMaxDelay))
+	if err != nil {
+		return nil, err
+	}
+	endpoint := &StreamEndpoint{
+		config:    config,
+		endpoints: endpoints,
+		target:    target,
+		conn:      conn,
+	}
+
+	return endpoint, nil
+}
+
+func (se *StreamEndpoint) Close() error {
+	return se.conn.Close()
+}
+
+func (se *StreamEndpoint) ConsumeStream(streamName string, opts ...ConsumerConfigOpt) *Consumer {
+
 	config := defaultConsumerConfig()
 	for _, opt := range opts {
 		opt(config)
 	}
 
 	ch := make(chan *stream.Event, config.BufferLen)
-	consumer := &Consumer{
+	c := &Consumer{
 		StreamName: streamName,
 		EvtChan:    ch,
 		config:     config,
-		endpoints:  endpoints,
-		target:     target,
 	}
+
+	var monitoringHolder = consumerMonitoring(streamName, se.endpoints)
+
 	go func() {
-		consumer.run()
+		for se.conn.GetState() != connectivity.Shutdown {
+			waitTillReady(se)
```

Comment:
We should also check inside waitTillReady that the connectivity isn't in SHUDOWN
Because a client can also close the connection

Reply:
yes